udp/server.go

```diff
@@ -0,0 +1,303 @@
+		blockwiseEnable:                opts.blockwiseEnable,
+		blockwiseTransferTimeout:       opts.blockwiseTransferTimeout,
+		multicastHandler:               client.NewHandlerContainer(),
+		multicastRequests:              &sync.Map{},
+		msgID:                          msgID,
+		serverStartedChan:              serverStartedChan,
+		onNewClientConn:                opts.onNewClientConn,
+		transmissionNStart:             opts.transmissionNStart,
+		transmissionAcknowledgeTimeout: opts.transmissionAcknowledgeTimeout,
+		transmissionMaxRetransmit:      opts.transmissionMaxRetransmit,
+
+		conns: make(map[string]*client.ClientConn),
+	}
+}
+
+func (s *Server) Serve(l *coapNet.UDPConn) error {
+	if s.blockwiseSZX > blockwise.SZX1024 {
+		return fmt.Errorf("invalid blockwiseSZX")
+	}
+
+	m := make([]byte, s.maxMessageSize)
+	s.listenMutex.Lock()
+	if s.listen != nil {
+		s.listenMutex.Unlock()
+		return fmt.Errorf("server already serve: %v", s.listen.LocalAddr().String())
+	}
+	s.listen = l
+	close(s.serverStartedChan)
+	s.listenMutex.Unlock()
+	defer func() {
+		s.closeSessions()
+		s.listenMutex.Lock()
+		defer s.listenMutex.Unlock()
+		s.listen = nil
+		s.serverStartedChan = make(chan struct{}, 1)
+	}()
+
+	var wg sync.WaitGroup
+	for {
+		buf := m
+		n, raddr, err := l.ReadWithContext(s.ctx, buf)
+		if err != nil {
+			wg.Wait()
+			select {
+			case <-s.ctx.Done():
+				return nil
+			default:
+				return err
+			}
+		}
+		buf = buf[:n]
+		cc, created := s.getOrCreateClientConn(l, raddr)
+		if created {
+			if s.onNewClientConn != nil {
+				s.onNewClientConn(cc)
+			}
+			if s.keepalive != nil {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					err := s.keepalive.Run(cc)
+					if err != nil {
+						s.errors(err)
+					}
+				}()
+			}
+		}
+		err = cc.Process(buf)
+		if err != nil {
+			cc.Close()
+			s.errors(err)
+		}
+	}
+}
+
+// Stop stops server without wait of ends Serve function.
+func (s *Server) Stop() {
+	s.cancel()
+	s.closeSessions()
+}
+
+func (s *Server) closeSessions() {
+	s.connsMutex.Lock()
+	tmp := s.conns
+	s.conns = make(map[string]*client.ClientConn)
+	s.connsMutex.Unlock()
+	for _, v := range tmp {
+		v.Close()
+	}
+}
+
+func (s *Server) conn() *coapNet.UDPConn {
+	s.listenMutex.Lock()
+	serverStartedChan := s.serverStartedChan
+	s.listenMutex.Unlock()
+	select {
+	case <-serverStartedChan:
+	case <-s.ctx.Done():
+	}
+	s.listenMutex.Lock()
+	defer s.listenMutex.Unlock()
+	return s.listen
+}
+
+func (s *Server) getOrCreateClientConn(UDPConn *coapNet.UDPConn, raddr *net.UDPAddr) (*client.ClientConn, bool) {
+	s.connsMutex.Lock()
+	defer s.connsMutex.Unlock()
+	key := raddr.String()
+	cc := s.conns[key]
+	created := false
+	if cc == nil {
+		created = true
+		var blockWise *blockwise.BlockWise
+		if s.blockwiseEnable {
+			blockWise = blockwise.NewBlockWise(func(ctx context.Context) blockwise.Message {
+				return pool.AcquireMessage(ctx)
+			}, func(m blockwise.Message) {
```

Comment:
Is this due to type-casting of the return value from a concrete type to an interface? Please could you tidy up this nested statement e.g. by preparing the parameters in local variables?